src/transformers/CJSImportTransformer.ts

```diff
@@ -798,14 +798,23 @@ export default class CJSImportTransformer extends Transformer {
       }
 
       const specifierInfo = getImportExportSpecifierInfo(this.tokens);
+
       while (this.tokens.currentIndex() < specifierInfo.endIndex) {
         this.tokens.removeToken();
       }
-      if (!specifierInfo.isType && !this.shouldElideExportedIdentifier(specifierInfo.leftName)) {
-        const localName = specifierInfo.leftName;
+
+      if (!specifierInfo.isType) {
         const exportedName = specifierInfo.rightName;
-        const newLocalName = this.importProcessor.getIdentifierReplacement(localName);
-        exportStatements.push(`exports.${exportedName} = ${newLocalName || localName};`);
+        if (exportedName === "default") {
+          this.hadDefaultExport = true;
+        } else {
+          this.hadNamedExport = true;
+        }
+        if (!this.shouldElideExportedIdentifier(specifierInfo.leftName)) {
```

Comment:
I think the `shouldElideExportedIdentifier` check should actually be kept with the previous `if` statement, so that implicitly-elided exports are skipped as well. Otherwise, the `module.exports = exports.default;` line will be incorrectly added for this code:
```ts
type T = number;
export {T as default};
```

Reply:
If we kept it, the unit test will fail, because the `shouldElideExportedIdentifier` returning value of `export { x as default } from './foo'` is `true`.

---

In your case, we can actually export both a value and a type as default at the same time (though with TS errors)

https://www.typescriptlang.org/play?#code/PTAEAEBcGcFoDsD2BjAFgU2QawFA8gJ4AO6oAKqALyjwCuAtgEboBOA3DugB5GIuSgA3uVABDaKAAm6AGajaAGwEBfDpx58B0uYoEBGIA

But the output will remain only one `export default` and no named export, so we can add `module.exports = exports.default`.

Reply:
> If we kept it, the unit test will fail

Good point, sorry for not trying that out. I think my case above is still valid, though, see [babel example](https://babeljs.io/repl#?browsers=node%206&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=C4TwDgpgBAKlC8UB2BXAtgIwgJwNwCgIAPMAe22CgG84BDAZygBMIAzWlAG2AF9cg&debug=false&forceAllTransforms=false&modules=commonjs&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env%2Cstage-2%2Cstage-3%2Ctypescript&prettier=false&targets=Node-13&version=7.22.8&externalPlugins=babel-plugin-add-module-exports%401.0.4&assumptions=%7B%7D). If the export is erased due to being a type, it shouldn't count when deciding whether or not to emit `module.exports = exports.default`.

This is actually very similar to #806 that I just implemented, but that fix was just for ESM output since CJS happened to already be working. The core issue, I think, is that `export {...}` and `export {...} from 'foo'` are handled by the same code path, but only the first one should erase type-only exported names. This didn't actually cause any correctness issues for CJS in normal export elision, but it does matter for this case of implementing `enableLegacyBabel5ModuleInterop`. Just like in #806, I think the fix is to look ahead to distinguish the two cases.

This PR certainly puts the feature in better shape, so let me merge it now, and I'll do a quick follow-up to refactor the code from #806 to handle the type case here I mentioned. Thanks again!

> we can actually export both a value and a type as default at the same time

Agreed that this is possible, but I think if `default` is *only* a type, then it should be ignored for the purposes of deciding whether to emit `module.exports = exports.default`. That's consistent with Babel's behavior linked above.